### PR TITLE
support hint in sharding-proxy

### LINKED ISSUE #2979: support hint in sharding-proxy
hint was supported by sharding-jdbc by thread-local. but it can't work in sharding-proxy.
we can extend [sctl](https://shardingsphere.apache.org/document/current/cn/manual/sharding-proxy/usage/#sctl-sharding-proxy-control-language) to support it in sharding-proxy. it looks like that 

- sctl:hint set MASTER_ONLY=true
- sctl:hint set DatabaseShardingValue=xx.     ## xx: sharding value
- sctl:hint addTableShardingValue xx=yy         ## xx: logicTableName，yy:sharding value
- sctl:hint addDatabaseShardingValue xx=yy. ## xx: logicTableName，yy:sharding value

sctl:hint command can be sent to proxy through JDBC, e.g. `connection.createStatement("stcl:hint set xxxx")`, after proxy received the request, it should be dispatched to `HintSCTLCommandHander` like `ShardingCTLSetBackendHandler`, then stored the hint content in current channel(`BackendConnection`)

### PR DESCRIPTION
**Pull request by betterjava**
Fixes #2979.

Changes proposed in this pull request:
- support hint in sharding-proxy,client can write sql like bellow:
  - addDatabaseShardingValue
    sctl:hint set hint_type=DATABASE_TABLES
    sctl:hint set addDatabaseShardingValue xx=yy ## xx:logicTable yy:shardingValue

  - addTableShardingValue
    sctl:hint set hint_type=DATABASE_TABLES
    sctl:hint set addTableShardingValue xx=yy ## xx:logicTable yy:shardingValue

  - setDatabaseShardingValue
    sctl:hint set hint_type=DATABASE_ONLY
    sctl:hint set setDatabaseShardingValue yy ## yy:shardingValue

  - setMasterRouteOnly
     sctl:hint set hint_type=MASTER_ONLY

  - clear 
    sctl:hint clear

### CODE REVIEW

**Comment by cherrylzhao** [comment]
good job! I'd like to review this PR today.

**Comment by cherrylzhao** [comment]
I found your editor setting is different with us, please check it at first
image

**Comment by betterjava** [comment]
ok,I will check.

**Comment by betterjava** [comment]
it is done, thanks for your reminder.

**Comment by betterjava** [comment]
it seems like some test failed in sharding-ui-distribution,how can I let the checks run again?

**Comment by cherrylzhao** [comment]
> it seems like some test failed in sharding-ui-distribution,how can I let the checks run again?

I'll contact the ui developer fix it as soon as possible

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    @Override
+    public BackendResponse execute() {
+        if (!supportHint) {
+            throw new UnsupportedOperationException(String.format("%s should be true, please check your config", ShardingPropertiesConstant.PROXY_HINT_ENABLED.getKey()));
+        }
+        Optional<ShardingCTLHintStatement> shardingTCLStatement = new ShardingCTLHintParser(sql).doParse();
+        if (!shardingTCLStatement.isPresent()) {
+            return new ErrorResponse(new InvalidShardingCTLFormatException(sql));
+        }
+        HintCommand hintCommand = shardingTCLStatement.get().getHintCommand();
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java]: you can try to use strategy design pattern here to make code more clear.
e.g. add HintCommandExecutor interface, use `HintCommandExecutorFacotry.newInstance` to refactor it.

**Comment by cherrylzhao** [comment]
> > it seems like some test failed in sharding-ui-distribution,how can I let the checks run again?
> 
> I'll contact the ui developer fix it as soon as possible

I have fixed it temporarily.

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.command.HintSetDatabaseShardingValueCommand;
+
+/**
+ * Sharding CTL hint backend handler.
+ *
+ * @author liya
+ */
+public final class ShardingCTLHintBackendHandler implements TextProtocolBackendHandler {
+
+    private static final ThreadLocal<HintManager> HINT_MANAGER_HOLDER = new ThreadLocal<>();
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java]: it is not necessary to add this HOLDER, `HINT_MANAGER_HOLDER` have existed in HintManager now.

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/HintType.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal;
+
+/**
+ * Hint type.
+ *
+ * @author liya
+ */
+public enum HintType {
+    DATABASE_TABLES, DATABASE_ONLY, MASTER_ONLY
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/HintType.java]: should add an empty line here.

File: sharding-proxy/sharding-proxy-backend/src/test/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import static org.mockito.Mockito.when;
+
+/**
+ * Test for ShardingCTLHintBackendHandler.
+ *
+ * @author liya
+ */
+public final class ShardingCTLHintBackendHandlerTest {
+
+    private BackendConnection backendConnection;
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/test/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandlerTest.java]: we support to add a mock property in this way
```java
@RunWith(MockitoJUnitRunner.class)
public final class ShardingCTLHintBackendHandlerTest {
    
    @Mock
    private BackendConnection backendConnection;

}

```

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintParser.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        setDatabaseShardingValueMatcher = Pattern.compile(setDatabaseShardingValueRegex, Pattern.CASE_INSENSITIVE).matcher(sql);
+        addDatabaseShardingValueMatcher = Pattern.compile(addDatabaseShardingValueRegex, Pattern.CASE_INSENSITIVE).matcher(sql);
+        addTableShardingValueMatcher = Pattern.compile(addTableShardingValueRegex, Pattern.CASE_INSENSITIVE).matcher(sql);
+        errorParameterMatcher = Pattern.compile(errorParameterRegex, Pattern.CASE_INSENSITIVE).matcher(sql);
+        clearMatcher = Pattern.compile(clearRegex, Pattern.CASE_INSENSITIVE).matcher(sql);
+    }
+
+    @Override
+    public Optional<ShardingCTLHintStatement> doParse() {
+        if (setMatcher.find()) {
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintParser.java]: select HintStatement logic can inline into HintCommandExecutorFacotry.

**Comment by cherrylzhao** [comment]
@betterjava 
how are things going on, do you need any help?

**Comment by betterjava** [comment]
@cherrylzhao,I am feel so sorry,I went home last week to handle private things.I will continue this today.

**Comment by betterjava** [comment]
Of course I will.I take them as opportunities to learn from you,thanks a lot.

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java (hunk 2 above)
Comment (betterjava) [reply sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java]: HintManager.getInstance() will throw an Exception if we invoke twice.
Should I refactor that like below?
```
    public static HintManager getInstance() {
        if(null == HINT_MANAGER_HOLDER.get()){
            HINT_MANAGER_HOLDER.set(new HintManager());
        }
        return HINT_MANAGER_HOLDER.get();
    }
```

Comment (cherrylzhao) [reply sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/ShardingCTLHintBackendHandler.java]: I have understood the problem, it's better to create new issue to consider about it in future.
let me cancel this review.

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/command/HintSetCommand.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.HintCommand;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.HintType;
+
+/**
+ * Hint Set command.
+ *
+ * <p>
+ * example: sctl:hint set hint_type=database_tables.
+ * hint_type: database_tables or database_only or database_only.
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/command/HintSetCommand.java]: have duplicate database_only, should master_only.

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/command/HintSetDatabaseShardingValueCommand.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.command;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.HintCommand;
+
+/**
+ * Hint SetDatabaseShardingValue command.
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/command/HintSetDatabaseShardingValueCommand.java]: should be `Hint set database sharding value command` or use {@code} instead

File: sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/executor/HintAddDatabaseShardingValueExecutor.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.shardingproxy.backend.response.BackendResponse;
+import org.apache.shardingsphere.shardingproxy.backend.response.update.UpdateResponse;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.HintCommandExecutor;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.HintManagerHolder;
+import org.apache.shardingsphere.shardingproxy.backend.text.sctl.hint.internal.command.HintAddDatabaseShardingValueCommand;
+
+/**
+ * Hint AddDatabaseShardingValue command executor.
```
Comment (cherrylzhao) [inline sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/text/sctl/hint/internal/executor/HintAddDatabaseShardingValueExecutor.java]: generated doc should be an effective word, or use `{@code}` for class or variable name.

**Comment by betterjava** [comment]
> it seems that hint_type is not necessary, do you agree with? we can only provide the following command?
> 
> * sctl:hint set MASTER_ONLY=true
> * sctl:hint set DatabaseShardingValue=xx
> * sctl:hint addTableShardingValue xx=yy
> * sctl:hint addDatabaseShardingValue xx=yy

Agree.
That seems more clearly.I will do some refactor.

**Pull request merged**